### DISCUSSION TITLE
Les objets communs

### DISCUSSION THREAD

**Discussion by Pebloop** [TempoDev/Arcade-Toulouse#10] [Ideas]
Yo !
Ici on parle des objets qu'on doit avoir en commun pour l'arcade, genre **sprite**, **sound**, **input**, **events**, ...
Il faut bien les définir pour que tout le monde puisse les interpréter depuis son *core* ^^'

**Comment by smbss1** [comment]
non je ne pense que l'on soit obliger d'avoir les objets en commun.

**Reply by Pebloop** [reply]
Si parce que tu ne peux pas coder des jeux intercompatible si tu ne définis pas ce que c'est.
Par exemple comment dans ton jeu tu crées un sprite ?
Sachant que ça doit fonctionner pour tout le monde.

**Comment by Pebloop** [comment]
Déja je pense qu'il faut séparer comment fonctionneront les objets des events.

* les objets : Les sprites, sons, etc seraient des classes/interfaces, c'est leurs existences qui, envoyé à la lib graphique, les fera afficher.
Pour ça, juste une interface commune serait suffisant.

* les events : les évenements inputs seraient des handlers, elles doivent être appelés et retourner une valeur qui va faire réagir le jeu différament.
Pour ceux là en revanche, il y a plusieurs solutions.

Mais la plus simple à laquelle on pense avec nasso, et de faire un callback, soit ajouter des fonction `onPressed`, `onClick`, `onreleased`, ... aux interfaces `IGame` et `IGraphic` (donc à `ILibrary`), que le core devra appeler lorsqu'il captera un event, et qui devra donc être définie dans le jeu.

> Pour ceux qui ne voient pas le délire, je vous invite à regarder comment fonctionne la [gestion d'événements sur android](https://www.geeksforgeeks.org/handling-click-events-button-android/), particulièrement la partie `Implementing View.OnClickListener in your Activity or fragment`, recopiée gentiment ci-dessous :

<Details>
<Summary>Implementing View.OnClickListener in your Activity or fragment</Summary>

To implement View.OnClickListener in your Activity or Fragment, you have to override onClick method on your class.

Firstly, link the button in xml layout to java by calling findViewById() method.

R.id.button_send refers the button in XML.

`mButton.setOnClickListener(this);` means that you want to assign listener for your Button “on this instance” this instance represents OnClickListener and for this reason your class have to implement that interface.

```xml
<RelativeLayout
	xmlns:android="http:// schemas.android.com/apk/res/android"
	xmlns:app="http:// schemas.android.com/apk/res-auto"
	xmlns:tools="http:// schemas.android.com/tools"
	android:layout_width="match_parent"
	android:layout_height="match_parent"
	tools:context="com.example.sample.MainActivity" >

	<Button
		android:layout_width="wrap_content"
		android:layout_height="wrap_content"
		android:id="@+id/button_send" />
</RelativeLayout>

```

MainActivity code:
```java
public class MainActivity extends AppCompatActivity
	implements View.OnClickListener {
	private Button mButton;

	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		mButton = findViewById(R.id.button_send);
		mButton.setOnClickListener(this);
	}

	@Override
	public void onClick(View view)
	{
		switch (view.getId()) {
		case R.id.button_send:
			// Do something
		}
	}
}

```
If you have more than one button click event, you can use switch case to identify which button is clicked.
</Details>

**Comment by Pebloop** [comment]
Il faut aussi définir spécifiquement quels seraient ces objets & events, pour les plus logiques je donne :

Objets ;
* Text
* Ascii (ASCII ONLY)
* Sprite (GRAPHIC ONLY)
* Sound
* (Custom) Property (?)

Events :
* Keyboard & Mouses
    * onClick
    * onPress
    * onRelease
* Mouse
    * position
* Exit

Je vois pas d'autres trucs comme ça de tête.

<details>
<Summary>Removed</Summary>

-  Shapes
    - Circle
    - Polygon (?)
    - Square
    - Line
</details>

**Reply by nasso** [reply]
Je sais pas si c'est nécessaire d'avoir tout les shapes ou les sons. Pour convenir au sujet seul Ascii et Sprite sont nécessaires (le jeu peut se débrouiller tout seul pour afficher du texte et des formes avec). A la limite le son peut être cool mais je pense que c'est plus un bonus, et il faut donc trouver un moyen de spécifier des objets custom qui seront ignorés par les implémentations qui ne les supportent pas.

**Reply by Pebloop** [reply]
> Je sais pas si c'est nécessaire d'avoir tout les shapes ou les sons. Pour convenir au sujet seul Ascii et Sprite sont nécessaires (le jeu peut se débrouiller tout seul pour afficher du texte et des formes avec). A la limite le son peut être cool mais je pense que c'est plus un bonus, et il faut donc trouver un moyen de spécifier des objets custom qui seront ignorés par les implémentations qui ne les supportent pas.

Ouais je remove shapes parce que ça peut être assez complexe pour pas grand chose, mais son ce serait bien de le laisser, ceux qui ne peuvent/veulent pas font juste une fonction vide et ceux qui veulent l'implémentent (surtout que ça n'est pas indispensable pour le jeu final).

**Comment by Pebloop** [comment]
En revanche il faut également penser à un système de **Layers**
Sois c'est également une classe commune qui contient les entitées de son layer, soit un composant qui serait juste un nombre, ...
Peut être une fonctions dans l'interface `IEntity` qui permet de set le layer ?

**Reply by nasso** [reply]
Une solution serai de définir la scène comme étant un conteneur de layer. Les entités sont alors contenues dans chacun de ces layers:
```cpp
#include <cstddef>

class IEntity {
    // ...
};

/* Entity container. */
class ILayer {
public:
    virtual ~ILayer() = default;

    virtual IEntity& createEntity() = 0;
    virtual void removeEntity(const IEntity&);
};

/* Layer container. */
class IScene {
public:
    virtual ~IScene() = default;

    virtual ILayer& addLayer(std::size_t index) = 0;
    virtual void moveLayer(const ILayer&, std::size_t new_index) = 0;
    virtual void removeLayer(const ILayer&) = 0;
};
```

Perso j'ai un avi mitigé sur cette approche, je préférerai avoir un mécanisme de "profondeur" par entité qui donne bien plus de libertés au jeu et simplifie grandement l'API. Le "bémol" est de devoir faire un tri côté lib graphique, mais c'est ce que font la plupart des jeux aujourd'hui de toute manière.

**Reply by Pebloop** [reply]
Du coup après discussion avec nasso, on pense qu'il vaut mieux assigner le layer sous forme de profondeur dans IEntity, avec genre `setDepth()` et `getDepth()`.

Si vous ne comprenez pas ou n'êtes pas d'accord, dites le maintenant.

**Reply by fluffyloafie** [reply]
C'est quoi un Layer en pratique ? Un ensemble d'entités ? Et comment on met en place de la profondeur ?

**Reply by Jad31** [reply]
Je n'ai pas compris la profondeur non plus

**Reply by nasso** [reply]
L'idée c'est juste de donner un ordre d'affichage au entités, pour par exemples s'assurer qu'une entité soit toujours dessinée "après" une autre, afin de paraître "devant".

**Reply by fluffyloafie** [reply]
Okéééééé j'ai compris ça. Et les layers du coup ? C'est dans le principe de profondeur ?

**Reply by nasso** [reply]
Oui c'est une manière de faire. Un peu comme des layers photoshops, tes entités sont répartie dans des "layers" triés qui permette de définir quelles entités s'affichent devant quelles autres entités.

**Reply by Pebloop** [reply]
> Okéééééé j'ai compris ça. Et les layers du coup ? C'est dans le principe de profondeur ?

En gros le principe de layer, comme dit par nasso c'est le fait de trier les entités par 'calques' comme dans photoshop.
Plusieurs feuilles de calques sont superposés et tu mets tes objets dans le bon calque pour qu'ils soient devant/derrière d'autres.
Le layer peut être créé techniquement en assignant l'id d'un layer à chaques entities (par exemple 0 le plus au dessus et 100 le plus en dessous), ou alors en faisant une liste d'entités pour chaques layers.

Le principe de depth, ou profondeur, et d'assigner un axe Z de profondeur aux entités (en plus de la position x, y) qui permet de définir la profondeur z d'un objet, et donc de savoir s'il est devant ou derrière un autre.

Ces deux notions sont 2 façon différentes d'aborder la profondeur en 2D mais au fond ce qui les differencies vraiment c'est que Pour les layers on va utiliser des nombres entier (1, 2, 3, ...) Tandis que l'autre méthode nécessite des nombres décimaux (0.6, 30.77, ...).

**Comment by nadmax** [comment]
Concernant les events, comment pourrait-on les représenter ?
Sous la forme d'une enum ? D'une classe ?

**Reply by nasso** [reply]
je vois bien une sorte de struct/classe par type d'event

```cpp
struct KeyEvent {
    enum {
        KEY_ESCAPE,
        KEY_RIGHT,
        KEY_LEFT,
        KEY_UP,
        KEY_DOWN,
        // ...
    } key;
    enum {
        KEY_PRESSED,
        KEY_RELEASED,
        KEY_REPEAT,
    } action;
    // ...
};

struct MouseEvent {
    int x;
    int y;
    enum {
        BTN_PRIMARY,
        BTN_SECONDARY,
        BTN_AUXILIARY,
    } button;
    // ...
};
```

ça permet d'avoir des prototypes plus consistants pour les events:

```cpp
class IGame {
    // ...
    virtual void onKeyDown(const KeyEvent&) = 0;
    virtual void onMouseDown(const MouseEvent&) = 0;
    // ...
};
```

De plus, j'en discutait avec @Jad31, @Arcahub et Kenny; ça permet plus facilement d'avoir derrière une implémentation plus générique qui utiliserai des templates:

```cpp
class IGraphics {
    // ...
    
    // Template sur le type d'événement
    // Le callback prend en paramètre une référence constante sur l'event
    template <typename Event>
    void on(std::function<void(const Event&)> callback)
    {
        // ...
    }
    
    // ...
};

void onKeyDown(const KeyEvent& e)
{
    // ...
}

// on s'abonne à un type d'événement:
IGraphics& my_graphlib = ...;
my_graphlib.on<KeyEvent>(onKeyDown);
```

Le problème après c'est que les templates vont pas forcément être faciles à mettre en place pour nos interfaces communes. Faudra voir avec Jerome, mais en soit on peut juste déclarer le prototype du template et ensuite le spécialiser chacun dans nos `.cpp`, pour `KeyEvent`, `MouseEvent`, etc...

**Reply by Pebloop** [reply]
Ce qui tu propose nasso me semble un peu compliqué, est-ce vraiment **nécessaire** de faire tout ce système d'abonnement quand tu peux juste créer une suite de méthodes :
```cpp
void onKeyDown(const Key &key);
void onKeyPressed(const Key &key);
void onKeyReleased(const Key &key);
void onMouseDown(const MouseEvent &mouse);
void onMousePressed(const MouseEvent &mouse);
void onMouseReleased(const MouseEvent &mouse);
```

Bon pour les enum je pense que de toute façon y a pas le choix, donc go partir sur un `event.hpp`
```cpp
enum {
    KEY_ESCAPE,
    KEY_BACKSPACE,
    KEY_RIGHT,
    KEY_LEFT,
    KEY_UP,
    KEY_DOWN,
    KEY_SHIFT_LEFT,
    KEY_SHIFT_RIGHT,
    KEY_CTRL_LEFT,
    KEY_CTRL_RIGHT,
    KEY_ALT_LEFT,
    KEY_ALT_RIGHT,
    KEY_TAB,
    KEY_PAGE_UP,
    KEY_PAGE_DOWN,
    KEY_DELETE,
    KEY_INSER,
    KEY_END,
    KEY_SPACE,
    KEY_F1,
    KEY_F2,
    KEY_F3,
    KEY_F4,
    KEY_F5,
    KEY_F6,
    KEY_F7,
    KEY_F8,
    KEY_F9,
    KEY_F10,
    KEY_F11,
    KEY_F12,
    KEY_A = 65,
    KEY_B = 66,
    KEY_C = 67,
    KEY_D = 68,
    KEY_E = 69,
    KEY_F = 70,
    KEY_G = 71,
    KEY_H = 72,
    KEY_I = 73,
    KEY_J = 74,
    KEY_K = 75,
    KEY_R = 76,
    KEY_L = 77,
    KEY_M = 78,
    KEY_N = 79,
    KEY_O = 80,
    KEY_P = 81,
    KEY_Q = 82,
    KEY_R = 83,
    KEY_S = 84,
    KEY_T = 85,
    KEY_U = 86,
    KEY_V = 87,
    KEY_W = 88,
    KEY_X = 89,
    KEY_Y = 90,
    KEY_Z = 91,
    KEY_1,
    KEY_2,
    KEY_3,
    KEY_4,
    KEY_5,
    KEY_6,
    KEY_7,
    KEY_8,
    KEY_9,
    KEY_0,
} Key;

struct MouseEvent {
    int x;
    int y;
    enum {
        MOUSE_PRIMARY,
        MOUSE_SECONDARY,
        MOUSE_AUXILIARY,
    } button;
};
```

**Reply by nasso** [reply]
Pour ce côté là oui, mais la partie abonnement j'y pensais plus pour permettre au core d'écouter les évènements générés par la librairie graphique.

**Reply by Pebloop** [reply]
Ouais je vois mais je pense que c'est trop flou pour être utilisé de manière générique ?

**Reply by Pebloop** [reply]
Enfin je veux dire, je sais pas si tout le monde comprend et voit comment l'implémenter + ça n'a pas l'air suffisament indispensable pour forcer tout le monde à faire un truc aussi compliqué.

**Reply by nasso** [reply]
Oui je présentait juste l'idée pour montrer ce qui est possible si on décide de partir sur un truc similaire.

**Reply by nasso** [reply]
Par contre je vois toujours pas ou ces méthodes vont se retrouver: il faut un moyen d'intercepter les events côté `IGame` et un moyen de les émettre côté `IGraphic`.

**Reply by Pebloop** [reply]
Bah
```cpp
void onKeyDown(const Key &key);
void onKeyPressed(const Key &key);
void onKeyReleased(const Key &key);
void onMouseDown(const MouseEvent &mouse);
void onMousePressed(const MouseEvent &mouse);
void onMouseReleased(const MouseEvent &mouse);
```
C'est `IGame`, et on peut imgaginer côté `IGraphic`
```cpp
std::vector<Key> getKeyDown() const;
std::vector<Key> getKeyPressed() const;
std::vector<Key> getKeyReleased() const;
std::vector<MouseEvent> getMouseDown() const;
std::vector<MouseEvent> getMousePressed() const;
std::vector<MouseEvent> getMouseReleased() const;
```
ouun truc du genre, proposez si vous trouvez mieux mdr

**Reply by nasso** [reply]
Une solution pour le côté `IGraphic` serai de passer par `IScene`:

````cpp
void update(IScene& scene)
{
    MouseEvent event;

    scene.pushEvent(event);
}
```

Derrière, l'implémentation de `IScene` côté core permettra d'appeler les bonnes méthodes sur `IGame`.

**Comment by Jad31** [comment]
Pour les interfaces IScene, ILayer etc.., on se met d'accord sur quoi et qu'est ce qu'on fait au final ?

**Reply by fenikssc** [reply]
Faudrait le push pour rendre la chose officielle (j'écris ça pour remettre le sujet en marche) @TempoDev

**Reply by Pebloop** [reply]
Bah y a pleins de gens qui font des pull requests et je passe un bon moment chaques jours à review et à merge, mais je peux pas être tout le temps la non plus, j'ai aussi des projets.

En ce qui concerne ce qui a été décidé, y a une partie encore en débat et une autre déja décidé avec des issues.
Ce qui manque c'est la doc, et j'aimerais bien m'y ateler quand j'aurais le temps.
Je sortirais une nouvelle release avec des fix dès que possible (dans la journée si j'y arrive)
